anonlink/similarities.py

```diff
@@ -0,0 +1,100 @@
+import collections
+import functools
+import heapq
+import itertools
+import operator
+from typing import Counter, DefaultDict, Dict, List, Optional, Sequence, Tuple
+
+import numpy as np
+
+from .typechecking import Record
+
+
+def _hamming_sim_f(clk1: Record, clk2: Record) -> float:
+    assert len(clk1) == len(clk2)
+    assert len(clk1) != 0
+    return sum(map(operator.eq, clk1, clk2)) / len(clk1)
+
+
+def _hamming_similarity_k(
+    datasets: Sequence[Sequence[Record]],
+    threshold: float,
+    k: int
+) -> Tuple[Sequence[Sequence[int]], Sequence[float]]:
+    # Dictionaries support sparsity, but these could also be list...
+    candidates = tuple(collections.defaultdict(list) for _ in datasets)  # type: Tuple[DefaultDict[int, List], ...]
+
+    for records in itertools.product(enumerate(datasets[0]),
+                                     enumerate(datasets[1])):
+        (i0, clk0), (i1, clk1) = records
+        sim = _hamming_sim_f(clk0, clk1)
+
+        if sim >= threshold:
+            c = sim, (i0, i1)
+            candidates[0][i0].append(c)
+            candidates[1][i1].append(c)
+
```

Comment:
This for loop is almost exactly the same in the two functions; can't you factor it out somehow?

Reply:
Good catch. Done!